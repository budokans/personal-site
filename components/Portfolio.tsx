import { SimpleGrid } from "@chakra-ui/layout";
import { useMediaQuery } from "@chakra-ui/react";
import { Box, Flex, Heading, Text } from "@chakra-ui/layout";
import { Img } from "@chakra-ui/image";
import { Button } from "@chakra-ui/button";
import { useEffect, useState } from "react";
import { useFeatureContext } from "../lib/featureContext";

interface PortFolioItemProps {
  idx: number;
}

interface PortfolioImageProps {
  src: string;
  alt: string;
}

const Portfolio: React.FC = ({ children }) => {
  return (
    <SimpleGrid
      columns={[1, 1, 2]}
      spacing={3}
      w={["full", "90%", "80%", "900px"]}
      mt={20}
      columnGap={10}
      rowGap={8}
    >
      {children}
    </SimpleGrid>
  );
};

const PortfolioItem: React.FC<PortFolioItemProps> = ({ idx, children }) => {
  const { openFeature } = useFeatureContext();

  const handleClick = () => {
    openFeature(idx);
  };

  return (
    <Box role="button" position="relative" onClick={handleClick}>
      <Flex role="group" direction="row">
        {children}
      </Flex>
    </Box>
  );
};

const PortfolioItemImage: React.FC<PortfolioImageProps> = ({ src, alt }) => {
  return (
    <Img
      src={src}
      alt={`${alt} icon`}
      w="64px"
      h="64px"
      objectFit="cover"
      rounded="xl"
      mr={[3, 4]}
    />
  );
};

const PortfolioItemInner: React.FC<PortFolioItemProps> = ({
  idx,
  children,
}) => {
  const [isLargerThan768] = useMediaQuery("(min-width: 768px)");
  const [afterStyle, setAfterStyle] = useState({});

  useEffect(() => {
    if (isLargerThan768 && idx > 1) {
      setAfterStyle({});
    } else if (idx > 2) {
      setAfterStyle({});
    } else {
      setAfterStyle({
        background: "gray.300",
        position: "absolute",
        content: `""`,
        bottom: "-17px",
        width: "100%",
        height: "1px",
        display: "block",
      });
    }
  }, [isLargerThan768]);

  return (
    <Flex position="relative" alignItems="center" _after={afterStyle}>
      <Box mr={4}>{children}</Box>
      <PortfolioViewButton />
    </Flex>
  );
};

const PortfolioItemTitle: React.FC = ({ children }) => {
  return (
    <Heading as="h3" fontSize="md" fontWeight="normal" lineHeight="tall">
      {children}
    </Heading>
  );
};

const PortfolioItemText: React.FC = ({ children }) => {
  return (
    <Text
      fontSize="clamp(12px, 10.8px + 0.25vw, 14px)"
      color="#595959"
      noOfLines={2}
    >
      {children}
    </Text>
  );
};

const PortfolioViewButton: React.FC = () => {
  return (
    <Button rounded="xl" fontSize="sm" w={8} px={3} h={7} minW={16}>
      View
    </Button>
  );
};

export {
  Portfolio,
  PortfolioItem,
  PortfolioItemImage,
  PortfolioItemInner,
  PortfolioItemTitle,
  PortfolioItemText,
};
