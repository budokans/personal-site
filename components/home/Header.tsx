import { ReactElement, useEffect, useState } from "react";
import {
  Heading,
  Stack,
  Wrap,
  WrapItem,
  Button,
  Tooltip,
  useClipboard,
  useMediaQuery,
} from "@chakra-ui/react";
import { ChildrenProps, Contact } from "../../types";

interface HeaderLinksProps {
  readonly contacts: readonly Contact[];
}

interface HeaderLinkProps {
  readonly href: string;
}

export const Header = ({ children }: ChildrenProps): ReactElement => {
  return (
    <Stack w={["full", "90%", "80%", "900px"]} spacing={5}>
      {children}
    </Stack>
  );
};

export const HeaderText = ({ children }: ChildrenProps): ReactElement => {
  return (
    <Heading
      fontSize="clamp(24px, calc(14.40px + 2.00vw), 40px)"
      fontWeight="normal"
      lineHeight={1.3}
      as="h1"
    >
      {children}
    </Heading>
  );
};

export const HeaderLocation = ({ children }: ChildrenProps): ReactElement => {
  return (
    <Heading
      fontSize="clamp(15px, 10px + 1vw, 20px)"
      fontWeight="normal"
      lineHeight={1.3}
      as="h2"
    >
      {children}
    </Heading>
  );
};

export const HeaderLinks = ({ contacts }: HeaderLinksProps): ReactElement => {
  const [isLargerThan930] = useMediaQuery("(min-width: 930px)");
  const [isLargeViewport, setIsLargeViewport] = useState(false);

  useEffect(() => {
    isLargerThan930 ? setIsLargeViewport(true) : setIsLargeViewport(false);
  }, [isLargerThan930]);

  return (
    <Wrap spacing="3">
      {contacts.map((contact, idx) => {
        if (isLargeViewport && contact.type === "Email") {
          return (
            <TooltipBtn href={contact.value} key={-1}>
              {contact.type}
            </TooltipBtn>
          );
        } else {
          return (
            <Link
              href={
                contact.type === "Email"
                  ? `mailto:${contact.value}`
                  : contact.value
              }
              key={idx}
            >
              {contact.type}
            </Link>
          );
        }
      })}
    </Wrap>
  );
};

export const Link = ({
  href,
  children,
}: HeaderLinkProps & ChildrenProps): ReactElement => {
  return (
    <WrapItem>
      <Button
        as="a"
        borderRadius="2xl"
        h="7"
        fontSize="sm"
        href={href}
        target="_blank"
        rel="noopener noreferrer"
        data-testid={href}
      >
        {children}
      </Button>
    </WrapItem>
  );
};

export const TooltipBtn = ({
  href,
  children,
}: HeaderLinkProps & ChildrenProps): ReactElement => {
  const { hasCopied, onCopy } = useClipboard(href);

  return (
    <WrapItem>
      <Tooltip
        placement="bottom"
        borderRadius="lg"
        py="1.5"
        closeDelay={1000}
        label={hasCopied ? "Copied!" : "Click to copy!"}
        aria-label="Copy email"
      >
        <Button
          borderRadius="2xl"
          h="7"
          fontSize="sm"
          onClick={onCopy}
          data-testid="Tooltip"
        >
          {children}
        </Button>
      </Tooltip>
    </WrapItem>
  );
};
