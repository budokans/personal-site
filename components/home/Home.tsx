import { Flex } from "@chakra-ui/layout";

const Home: React.FC<{ blur: boolean }> = ({ blur, children }) => {
  return (
    <Flex
      direction="column"
      justifyContent={["space-between", "center"]}
      alignItems="center"
      minHeight="100vh"
      py={8}
      px={4}
      maxW={["100%", "930px"]}
      marginY="0"
      marginX="auto"
      sx={
        blur
          ? {
              filter: "blur(5px)",
              transition: "filter 0.5s ease-out 0.5s",
            }
          : { filter: "blur(0px)", transition: "filter" }
      }
    >
      {children}
    </Flex>
  );
};

export { Home };
