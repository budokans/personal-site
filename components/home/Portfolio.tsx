import { useEffect, useState } from "react";
import {
  Img,
  SimpleGrid,
  Box,
  Flex,
  Heading,
  Text,
  Button,
  useMediaQuery,
} from "@chakra-ui/react";
import { ImageProps } from "../../interfaces";
import { getBottomRowCount } from "../../lib/getBottomRowCount";

interface PortFolioItemProps {
  onPortfolioClick: () => void;
}

interface Compound {
  Item: React.FC<PortFolioItemProps>;
  Image: React.FC<ImageProps>;
  Inner: React.FC<{ idx: number; projectsCount: number }>;
  Title: React.FC;
  Text: React.FC;
  Button: React.FC;
}

type PortfolioCC = Compound & React.FC;

export const Portfolio: PortfolioCC = ({ children }) => {
  return (
    <SimpleGrid
      columns={[1, 1, 2]}
      w={["full", "90%", "80%", "900px"]}
      mt={20}
      columnGap={10}
      rowGap={4}
    >
      {children}
    </SimpleGrid>
  );
};

Portfolio.Item = ({ onPortfolioClick, children }) => {
  return (
    <Box
      role="button"
      position="relative"
      onClick={onPortfolioClick}
      data-testid="portfolio-item"
    >
      <Flex role="group" direction="row">
        {children}
      </Flex>
    </Box>
  );
};

Portfolio.Image = ({ src, alt }) => {
  return (
    <Img
      src={src}
      alt={`${alt} icon`}
      boxSize="64px"
      objectFit="cover"
      rounded="xl"
      mr={[3, 4]}
    />
  );
};

Portfolio.Inner = ({ idx, projectsCount, children }) => {
  const [isLargerThan768] = useMediaQuery("(min-width: 768px)");
  const [columnsCount, setColumnCount] = useState(1);
  const [renderBottomBorder, setRenderBottomBorder] = useState(true);
  const projectNumber = idx + 1;
  const borderStyle = "1px solid lightgrey";

  useEffect(() => {
    if (isLargerThan768) {
      setColumnCount(2);
    } else {
      setColumnCount(1);
    }
  }, [isLargerThan768]);

  const bottomRowCount = getBottomRowCount(projectsCount, columnsCount);

  useEffect(() => {
    if (projectNumber <= columnsCount) {
      setRenderBottomBorder(true);
    } else if (projectNumber > projectsCount - bottomRowCount) {
      setRenderBottomBorder(false);
    } else {
      setRenderBottomBorder(true);
    }
  }, [columnsCount]);

  return (
    <Flex
      position="relative"
      alignItems="center"
      borderBottom={renderBottomBorder ? borderStyle : ""}
      paddingBottom={4}
    >
      <Box mr={4}>{children}</Box>
      <Portfolio.Button />
    </Flex>
  );
};

Portfolio.Title = ({ children }) => {
  return (
    <Heading as="h3" fontSize="md" fontWeight="normal" lineHeight="tall">
      {children}
    </Heading>
  );
};

Portfolio.Text = ({ children }) => {
  return (
    <Text
      fontSize="clamp(12px, 10.8px + 0.25vw, 14px)"
      color="#595959"
      noOfLines={2}
    >
      {children}
    </Text>
  );
};

Portfolio.Button = () => {
  return (
    <Button rounded="xl" fontSize="sm" w={8} px={3} h={7} minW={16}>
      View
    </Button>
  );
};
