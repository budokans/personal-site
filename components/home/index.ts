export * as Header from "./Header";
export * from "./Home";
export * as Portfolio from "./Portfolio";
