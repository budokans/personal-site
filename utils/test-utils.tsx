import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
import { ChildrenProps } from "types";
import { ReactElement } from "react";

const withProviders = ({ children }: ChildrenProps): ReactElement => {
  return <ChakraProvider>{children}</ChakraProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult => render(ui, { wrapper: withProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
