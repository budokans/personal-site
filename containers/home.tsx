import { ReactElement } from "react";
import { SiteMetadata, Project } from "../types";
import { Header, Home, Portfolio } from "../components/home";
import { TextLink } from "../components/Link";

interface HomeContainerProps {
  readonly projects: readonly Project[];
  readonly metadata: SiteMetadata;
  readonly onPortfolioClick: (id: string) => void;
}

export const HomeContainer = ({
  projects,
  metadata,
  onPortfolioClick,
}: HomeContainerProps): ReactElement => {
  return (
    <Home.Home>
      <Header.Header>
        <Header.Heading>
          Hi, I&apos;m Steven. I&apos;m a full-stack developer who&apos;s helped
          build <TextLink url="https://rico.nz">Rico</TextLink> and{" "}
          <TextLink url="https://www.pivot-pointe.com">Pivot Pointe</TextLink>.
          I&apos;d love to help you with your software goals. Get in touch!
        </Header.Heading>

        <Header.HeaderContact>
          <Header.HeaderLocation>üìç {metadata.location}</Header.HeaderLocation>
          <Header.HeaderLinks contacts={metadata.contacts} />
        </Header.HeaderContact>
      </Header.Header>

      <Portfolio.Portfolio>
        {projects.map((project, idx) => {
          return (
            <Portfolio.PortfolioItem
              onPortfolioClick={onPortfolioClick}
              id={project.id}
              key={idx}
            >
              <Portfolio.PortfolioImage
                src={project.icon}
                alt={project.title}
              />
              <Portfolio.PortfolioInner
                idx={idx}
                projectsCount={projects.length}
              >
                <Portfolio.PortfolioTitle>
                  {project.title}
                </Portfolio.PortfolioTitle>
                <Portfolio.PortfolioText>
                  {project.shortBlurb}
                </Portfolio.PortfolioText>
              </Portfolio.PortfolioInner>
            </Portfolio.PortfolioItem>
          );
        })}
      </Portfolio.Portfolio>
    </Home.Home>
  );
};
