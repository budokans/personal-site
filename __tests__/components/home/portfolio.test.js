import { render, fireEvent } from "test-utils";
import { Portfolio } from "../../../components/home/Portfolio";

describe("<Portfolio />", () => {
  test("<Portfolio.Item /> renders a button and calls openFeature() on click", () => {
    const openFeature = jest.fn();
    const { getByRole, container } = render(
      <Portfolio.Item onPortfolioClick={openFeature}>Children</Portfolio.Item>
    );

    expect(getByRole("button")).toBeInTheDocument();
    fireEvent.click(getByRole("button"));
    expect(openFeature).toHaveBeenCalled();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Porfolio.Image /> renders an image with populated data from props", () => {
    const project = {
      title: "Gretsch Geeks",
      icon: "/images/gretsch-logo.jpg",
    };

    const { getByAltText, container } = render(
      <Portfolio.Image src={project.icon} alt={project.title} />
    );

    expect(getByAltText(/Gretsch Geeks icon/i)).toBeInTheDocument();
    expect(container.querySelector("img").getAttribute("src")).toBe(
      "/images/gretsch-logo.jpg"
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Title /> renders with populated data from props", () => {
    const project = {
      title: "Gretsch Geeks",
    };

    const { getByText, container } = render(
      <Portfolio.Title>{project.title}</Portfolio.Title>
    );

    expect(getByText(/Gretsch Geeks/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Text /> renders with populated data from props", () => {
    const project = {
      shortBlurb:
        "An example E-commerce buy-sell platform built with modern React, Apollo 3 and Keystone.js as a headless CMS.",
    };

    const { getByText, container } = render(
      <Portfolio.Text>{project.shortBlurb}</Portfolio.Text>
    );

    expect(getByText(/Apollo 3/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("<Portfolio.Button /> renders a button with text", () => {
    const { getByRole, getByText, container } = render(<Portfolio.Button />);

    expect(getByRole("button")).toBeInTheDocument();
    expect(getByText(/View/i)).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });
});
