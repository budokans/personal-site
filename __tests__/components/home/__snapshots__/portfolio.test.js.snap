// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portfolio /> <Porfolio.Image /> renders an image with populated data from props 1`] = `
<img
  alt="Gretsch Geeks icon"
  class="chakra-image css-1dzvt9d"
  src="/images/gretsch-logo.jpg"
/>
`;

exports[`<Portfolio /> <Portfolio.Button /> renders a button with text 1`] = `
<button
  class="chakra-button css-1a5ti90"
  type="button"
>
  View
</button>
`;

exports[`<Portfolio /> <Portfolio.Item /> renders a button and calls openFeature() on click 1`] = `
<div
  class="css-79elbk"
  role="button"
>
  <div
    class="css-1xhj18k"
    role="group"
  >
    Children
  </div>
</div>
`;

exports[`<Portfolio /> <Portfolio.Text /> renders with populated data from props 1`] = `
<p
  class="chakra-text css-1lzpwia"
>
  An example E-commerce buy-sell platform built with modern React, Apollo 3 and Keystone.js as a headless CMS.
</p>
`;

exports[`<Portfolio /> <Portfolio.Title /> renders with populated data from props 1`] = `
<h3
  class="chakra-heading css-1l30lph"
>
  Gretsch Geeks
</h3>
`;
